File: androidApp/src/main/java/de/simon/covid19/android/viewModels/HomeViewModel.kt
package de.simon.covid19.android.viewModels

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import de.simon.covid19.android.viewModels.actions.HomeAction
import de.simon.covid19.viewModels.states.HomeState
import de.simon.covid19.models.CountrySummary
import de.simon.covid19.repository.Repository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.*

class HomeViewModel(private val repository: Repository) : ViewModel() {

    companion object {
        private val TAG = HomeViewModel::class.simpleName
    }

    private val _viewState: MutableStateFlow<HomeState> = MutableStateFlow(HomeState.LOADING)
    val viewState = _viewState.asStateFlow()

    private val allCountries = mutableListOf<CountrySummary>()

    init {
        viewModelScope.launch {
            val summary = repository.getCovid19Summary()
            if (summary.isEmpty) {
                _viewState.value = HomeState.FAILED
            } else {
                allCountries.addAll(summary.countries)
                _viewState.value =
                    HomeState(
                        loading = false,
                        failed = false,
                        globalSummary = summary.global,
                        filteredCountries = allCountries,
                        input = ""
                    )
            }
        }
    }

    fun onAction(action: HomeAction) {
        Log.d(TAG, "onAction: ${action::class.simpleName}")
        when (action) {
            is HomeAction.InputChanged -> updateInput(action.input)
            is HomeAction.InputDelete -> updateInput("")
        }
    }

    private fun updateInput(input: String) {
        _viewState.value = _viewState.value.copy(
            input = input,
            filteredCountries = allCountries.filter {
                it.country.lowercase(Locale.ROOT).contains(input.lowercase(Locale.ROOT))
            })
    }
}

File: iosApp/iosApp/viewModels/HomeViewModel.swift
//
//  HomeViewModel.swift
//  iosApp
//
//  Created by Simon Lehmann on 16.10.21.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import shared

class HomeViewModel: ObservableObject {
    @Published var state: HomeState
    @Published var allCountries = [CountrySummary]()
    @Published var dateString: String
    
    private let repository: Repository
    
    init(repository: Repository) {
        self.repository = repository
        state = HomeState.companion.LOADING
        dateString = ""
    }
    
    func fetchCountries() {
        repository.getCovid19Summary { data, error in
            if let summary = data {
                self.allCountries = summary.countries
                self.state = HomeState(type: StateType.succeeded, globalSummary: summary.global, filteredCountries: self.allCountries, input: "")
                
                // Transform dateInstant to dateString
                if let dateInstant = summary.dateInstant {
                    let formatter = DateFormatter()
                    formatter.dateStyle = .medium
                    formatter.timeStyle = .medium
                    let date = Date(timeIntervalSince1970: TimeInterval(dateInstant.epochSeconds))
                    self.dateString = formatter.string(from: date)
                }
            }
            if let _ = error {
                self.state = HomeState.companion.FAILED
            }
        }
    }
    
    func onAction(action: HomeAction) {
        if let inputChanged = action as? HomeAction.InputChanged {
            // After InputDelete InputChanged is called again -> Prevent call again updateInput()
            if state.input != inputChanged.input {
                updateInput(input: inputChanged.input)
            }
        } else if action is HomeAction.InputDeleted {
            updateInput(input: "")
        }
    }
    
    private func updateInput(input: String) {
        if(state.input == input) {
            return
        }
        var filteredCountries = self.allCountries;
        if(!input.isEmpty) {
            filteredCountries = self.allCountries.filter { country in
                let countryName = country.country.lowercased()
                let contains = countryName.contains(input.lowercased())
                return contains
            }
        }
        state = HomeState(
            type: state.type,
            globalSummary: state.globalSummary,
            filteredCountries: filteredCountries,
            input: input)
    }
}
